pkg/aws/cloudfront/cloudfront.go

```diff
@@ -0,0 +1,140 @@
+package cloudfront
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	cf "github.com/aws/aws-sdk-go/service/cloudfront"
+)
+
+func NewClient(opts ...Option) (*Client, error) {
+	client := &Client{
+		awsCfg:  &aws.Config{},
+		timeout: 30 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	// By default, if no StaticCredentials are provided, NewSession will use environment variables
+	// AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN
+	session, err := session.NewSession(client.awsCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	client.cfApi = cf.New(session)
+
+	return client, nil
+}
+
+// Returns the current UTC time formatted using the format string
+//	"20060102150405"
+// which means: "2006-01-02 15:04:05"
+func currentTimeSeconds() string {
+	t := time.Now().UTC()
+	return t.Format("20060102150405")
+}
+
+// Verifies the Invalidation struct's pointers are non-nil
+func checkInvalidationStruct(invalidation *cf.Invalidation) error {
+	if invalidation == nil {
+		return fmt.Errorf("cloudfront Invalidation struct set to nil")
+	}
+
+	if invalidation.CreateTime == nil {
+		return fmt.Errorf("cloudfront Invalidation.CreateTime set to nil")
+	}
+
+	if invalidation.Id == nil {
+		return fmt.Errorf("cloudfront Invalidation.Id set to nil")
+	}
+
+	if invalidation.InvalidationBatch == nil {
+		return fmt.Errorf("cloudfront Invalidation.InvalidationBatch set to nil")
+	}
+
+	if invalidation.InvalidationBatch.CallerReference == nil {
+		return fmt.Errorf("cloudfront Invalidation.InvalidationBatch.CallerReference set to nil")
+	}
+
+	if invalidation.InvalidationBatch.Paths == nil {
+		return fmt.Errorf("cloudfront Invalidation.InvalidationBatch.Paths set to nil")
+	}
+
+	if invalidation.InvalidationBatch.Paths.Items == nil {
+		return fmt.Errorf("cloudfront Invalidation.InvalidationBatch.Paths.Items set to nil")
+	}
+
+	if invalidation.Status == nil {
+		return fmt.Errorf("cloudfront Invalidation.Status set to nil")
+	}
+
+	return nil
+}
+
+// Creates an Invalidation request
+func (c *Client) CreateInvalidation(distributionId string, paths []string) (*CreateInvalidationOutput, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+
+	output, err := c.cfApi.CreateInvalidationWithContext(ctx, &cf.CreateInvalidationInput{
+		DistributionId: aws.String(distributionId),
+		InvalidationBatch: &cf.InvalidationBatch{
+			// Using CallerReference as unique identifier for Invalidation request.
+			// Effectively rate limits CreateInvalidation requests for the same (Distribution_Id, Paths)
+			// to once per second.
+			CallerReference: aws.String(currentTimeSeconds()),
```

Comment:
The [CallerReference](https://docs.aws.amazon.com/sdk-for-go/api/service/cloudfront/#InvalidationBatch) is used as an unique identifier per CreateInvalidation request.

If a second request with the same CallerReference, distributionId and paths is made, it will be ignored by Cloudfront. If the CallerReference is the same but rest is different, Cloudfront will return error.

Here I'm using timestamp with seconds resolution, effectively rate limiting the requests for a given distributionId and paths to once per second.

Is this okay?

Is there a better use of CallerReference instead to take in user POST CSRF token and hash it with string/paths? This would guarantee a unique request but user would need to refresh their page to send in a second request.

Reply:
> Is there a better use of CallerReference instead to take in user POST CSRF token and hash it with string/paths

That is a nice approach for protecting the form from accidental invalidations if the button is clicked more than once. If the invalidation submission is done sync with the form submission, you can move the user to a results page on submission so they would have to reload the form to submit a new invalidation. 

Using a timestamp for the API end point seems reasonable but the response from the api would need to be clear if a single invalidation happens from multiple calls to the api. I am not positive how long the upstream api takes to reply. I presume this is all sync.

Reply:
The upstream api takes a few seconds to reply on Create, in my testing on our empty sandbox CDN.

I will leave the CallerReference as is for now, for simplicity. Maybe once the endpoints and UI is implemented can revisit and improve if needed.